cli/main.c

```diff
@@ -23,157 +24,55 @@
 #include <sys/stat.h>
 #include <unistd.h>
 
+#include "commands.h"
 #include "glome.h"
 
-const char *kUsage =
+#define GLOME_CLI_MAX_MESSAGE_LENGTH 4095
+
+static const char *kUsage =
     "Usage: \n"
-    "  To generate a new keypair (PRIVATE-KEY will be created if needed)\n"
-    "    %s PRIVATE-KEY >PUBLIC-KEY\n\n"
-    "  To generate a tag (defaults: MESSAGE:='' counter:=0):\n"
-    "    %s PRIVATE-KEY PEER-KEY [MESSAGE [COUNTER]]\n\n"
+    "  To generate a new keypair\n"
+    "    umask 077\n"
+    "    %s genkey | tee PRIVATE-KEY | %s pubkey >PUBLIC-KEY\n\n"
+    "  To generate a tag:\n"
+    "    %s tag --key PRIVATE-KEY --peer PEER-KEY [--counter COUNTER] "
+    "<MESSAGE\n\n"
     "  To verify a tag:\n"
-    "    %s PRIVATE-KEY PEER-KEY MESSAGE COUNTER TAG\n";
-
-static int decode_hex(uint8_t *dst, size_t dst_len, const char *in) {
-  size_t len = strlen(in);
-  if (len > 2 && in[0] == '0' && in[1] == 'x') {
-    len -= 2;
-    in += 2;
-  }
-  if (len > dst_len * 2 || len % 2 != 0) {
-    return -1;
-  }
-  size_t i;
-  for (i = 0; i < len / 2; i++) {
-    if (sscanf(in + (i * 2), "%02hhX", dst + i) != 1) {
-      fprintf(stderr, "ERROR while parsing byte %zu ('%c%c') as hex\n", i,
-              in[2 * i], in[2 * i + 1]);
-      return -3;
-    }
-  }
-  return i;
-}
+    "    %s verify --key PRIVATE-KEY --peer PEER-KEY --tag TAG [--counter "
+    "COUNTER] <MESSAGE\n";
 
-int read_key(const char *fname, uint8_t *buf, const size_t buf_len) {
-  int fd = open(fname, O_RDONLY);
-  if (fd == -1) {
-    if (errno != ENOENT) {
-      return -1;
-    }
-    if (fname[0] != '\0' && fname[1] == '-' && fname[1] == '\0') {
-      // '-' means stdin
-      fd = STDIN_FILENO;
-    } else {
-      // Try parsing the provided filename as the key
-      return decode_hex(buf, buf_len, fname);
-    }
-  }
-  return read(fd, buf, buf_len);
-}
-
-void print_hex(FILE *stream, const char *prefix, uint8_t *buf, size_t len) {
-  if (prefix != NULL) {
-    fputs(prefix, stream);
-  }
-  for (size_t i = 0; i < len; i++) {
-    fprintf(stream, "%02x", buf[i]);
-  }
-  fputs("\n", stream);
+static int print_help(int argc, char **argv) {
+  fprintf(stderr, kUsage, argv[0], argv[0], argv[0], argv[0]);
+  return EXIT_SUCCESS;
 }
 
-int main(int argc, char **argv) {
-  uint8_t private_key[GLOME_MAX_PRIVATE_KEY_LENGTH] = {0};
-  uint8_t public_key[GLOME_MAX_PUBLIC_KEY_LENGTH] = {0};
-  uint8_t peer_key[GLOME_MAX_PUBLIC_KEY_LENGTH] = {0};
-  uint8_t tag[GLOME_MAX_TAG_LENGTH] = {0};
-  uint8_t expected_tag[GLOME_MAX_TAG_LENGTH] = {0};
-  size_t expected_tag_len = 0;
-  char *message = NULL;
-  char *endptr;
-  bool verification = false;
-  unsigned long counter = 0;
+typedef int (*mainfunc)(int, char **);
```

Comment:
Would it be legit to name parameters in a func typedef?

Reply:
Yes, the rules seem to be the same as for function declarations - added names.